### PR TITLE
Add additional tests for replace member

### PR DESCRIPTION
**Pull request by sanebay**
Add OneMemberRestart , LeaderReplace, OneMemberDown test case.
Comments added in start of unit test

### CODE REVIEW

File: src/tests/test_raft_repl_dev_dynamic.cpp [hunk 1]
```diff
@@ -110,28 +116,176 @@ TEST_F(ReplDevDynamicTest, TwoMemberDown) {
...
+    g_helper->sync_for_test_start(num_members);
+
+    this->shutdown_replica(2);
+    LOGINFO("Shutdown replica 2");
+
+    std::this_thread::sleep_for(std::chrono::seconds(3));
+    if (g_helper->replica_num() == 0) {
+        // With existing raft repl dev group, write IO's, validate and call replace_member on leader.
+        LOGINFO("Writing on leader num_io={} replica={}", num_io_entries, g_helper->replica_num());
+        this->write_on_leader(num_io_entries, true /* wait_for_commit */);
```
Comment (xiaoxichen) [inline src/tests/test_raft_repl_dev_dynamic.cpp:164]: I think the `write_on_leader` needs to be execute on r0-r2.
In current implementation  R1/R2 skips the if-clause from L161-L171 and wait on L173.  If R1 is laggy (only commit on R0 and R3),   it is possible the R1 failed during verify

Comment (JacksonYao287) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: agree. in write_on_leader , the real leader will schedule real io, others will wait_for_commit.

File: src/tests/test_raft_repl_dev_dynamic.cpp [hunk 2]
```diff
@@ -110,28 +116,176 @@ TEST_F(ReplDevDynamicTest, TwoMemberDown) {
...
+
+    if (g_helper->replica_num() != member_in) {
+        LOGINFO("Writing on leader num_io={} replica={}", num_io_entries, g_helper->replica_num());
+        // With existing raft repl dev group, write IO's, validate and call replace_member on leader.
+        this->write_on_leader(num_io_entries, true /* wait_for_commit */);
+
+        // Leader will return error NOT_LEADER and yield leadership, sleep and connect again
+        // to the new leader.
+        LOGINFO("Replace old leader");
+        replace_member(db, g_helper->replica_id(member_out), g_helper->replica_id(member_in), 0,
```
[the remark this answers is not in the document]
Comment (sanebay) [reply src/tests/test_raft_repl_dev_dynamic.cpp:218]: But why client needs to know ? One option was to add additional error code which gateway needs to handle. Or send retry but gateway might retry same SM. Here the leader yield leadership and returns not leader and gateway retries by refreshing PG map.

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 1 above)
Comment (sanebay) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: write_on_leader is nop on other replica's. replica 0 is the leader and we are calling replace member with run_on_leader. I havent observed any replica's lagging in any of this or the original raft repl tests.

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxichen) [reply src/tests/test_raft_repl_dev_dynamic.cpp:218]: I am not talking about GW but CM,  when CM send a PG_MOVE request and get back a NOT_LEADER, should it take it as success or fail?

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 1 above)
Comment (xiaoxichen) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: >write_on_leader is nop on other replica's. 
It is not that "nop" but waiting for IOs to finish before proceeding.

I think all of the UTs except this one follows the pattern.  For example this:
https://github.com/eBay/HomeStore/pull/574/files#diff-8d61e3935e3a521ab097bbc27963ca7e583d90d6a597d1301a0378fa2cf33854R213

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sanebay) [reply src/tests/test_raft_repl_dev_dynamic.cpp:218]: Created an issue for CM to track this.

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 1 above)
Comment (sanebay) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: The reason is we are shutting down replica 2 and it doesnt db instances and write on leader will crash at line auto leader_uuid = dbs_[0]->repl_dev()->get_leader_id();

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxichen) [reply src/tests/test_raft_repl_dev_dynamic.cpp:218]: why it is a CM ticket instead that we send another error code?

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 1 above)
Comment (xiaoxichen) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: yeah that makes sense, 
```if (g_helper->replica_num() != 2) ```
if this is better?

Comment (JacksonYao287) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: >leader will crash at line auto leader_uuid = dbs_[0]->repl_dev()->get_leader_id();

can we add a check like this `if (is_replica_num_in({0, 1, member_in}))` in write_on_leader

**Review by JacksonYao287 — COMMENTED** [review]
generally, LGTM!

File: src/tests/test_raft_repl_dev_dynamic.cpp (hunk 1 above)
Comment (sanebay) [reply src/tests/test_raft_repl_dev_dynamic.cpp:164]: write_on_leader is generic and this case is specific for stop .

**Review by xiaoxichen — APPROVED** [review]

**Comment by xiaoxichen** [comment]
i still dont think skipping `wait_for_commit()` on other leader is right....but as it is an UT,  we can revisit if we see it flips.

**Review by JacksonYao287 — APPROVED** [review]
LGTM！

**Pull request merged**